source/common/ssl/context_impl.cc

```diff
@@ -469,12 +480,79 @@ ServerContextImpl::ServerContextImpl(ContextManagerImpl& parent, Stats::Scope& s
     RELEASE_ASSERT(rc == 1);
   }
 
+  // Hash configured SNIs for this context, so that sessions cannot be resumed across different
```

Comment:
I don't totally follow this. Should this be "so that sessions cannot be resumed across listeners with different configurations?" What is the downside of allowing resumption if the SNI cert matches?

Reply:
Yes, except `s/listeners/filter chains/`.

Imagine certificate for `*.example.com` that's used on different filter chains (e.g. `www.example.com` and `admin.example.com`). There is no reason sessions on those logically separate "virtual hosts" should be resumable simply because they use the same wildcard certificate.

Reply:
I think maybe we are using different lingo or I'm just confused. I think by far the most common use of SNI is going to be a single filter chain with multiple certificates, where host/authority header allows the virtual host in the route table to be selected. Very few people would regularly configure a full HTTP connection manager per SNI host I think. AFAICT the below check basically says that we don't allow resumption unless the filter chain has all the same server names that it had before. This seems wrong to me?

Reply:
Well, but that's not what the [data-plane-api] allows you to do, really.

Per current (and frozen) specification, `sni_domains` are part of the `FilterChainMatch`, which is used to select `FilterChain`. Each `FilterChain` contains a single `DownstreamTlsContext`, therefore multiple `DownstreamTlsContext` require multiple filter chains, and there is no other SNI to TLS certificate mapping.

cc @htuch 
 
[data-plane-api]: https://github.com/envoyproxy/data-plane-api

Reply:
OK I see. I think I totally missed that part of how we did the protos. I'm concerned users will find that burdensome since it will potentially require a different route/RDS table for each sever name. I now understand why you are calling this a hack w/o full filter chain matching support.

My concern here is that once this code goes out we are actually providing the behavior that I originally thought people would want (multiple server names -> 1 connection manager). Can we really take it away once we give it to them? At minimum we would have to put it through the full deprecation cycle.

This is a tough one. I'm wondering if it's worth it for me to help out and try to get the multiple filter chains working?

Reply:
@PiotrSikora I talked offline with @htuch and he had the idea that to help with back-compat, we should check that all the filter chains are identical and throw if they aren't. Then in the future we can relax this. I think that will make me OK with this.

I also now think I understand why the multiple filter chain stuff is complicated. You basically have to complete the handshake before you do any filter chain processing/selection. I can have a look at all of this offline and come back with some implementation notes for the future if that would be useful.

Reply:
FWIW, I was planning to use multiple filter chains heavily.  That decision isn't set in stone, but it's I was planning to model my configuration.

Reply:
@ggreenway we are going to implement multiple filter chains, it's just a matter of when. It's pretty urgent that we get base SNI support merged so I think this approach is fine to start with as long as we future proof it. Then I will likely look at the multiple filter chain aspect after this merges.

Reply:
sgtm